File: backend/app/api/rollouts.py
from __future__ import annotations

from fastapi import APIRouter

from app.services.rollouts_service import get_rollout_detail

router = APIRouter()


@router.get("/api/rollouts/{rollout_id}")
def get_rollout_api(rollout_id: str, stage_id: int, phase: str, source: str) -> dict:
    try:
        return get_rollout_detail(stage_id=stage_id, phase=phase, source=source, rollout_id=rollout_id)
    except ValueError as e:
        return {"ok": False, "error": str(e)}
    except Exception as e:
        return {"ok": False, "error": f"read rollout failed: {e}"}

File: frontend/src/api/rollouts.js
const DEFAULT_API_BASE = "";

function apiBase() {
  const base = import.meta.env.VITE_API_BASE || DEFAULT_API_BASE;
  return String(base).replace(/\/$/, "");
}

export async function fetchRollout({ stageId, phase, source, rolloutId, signal }) {
  const qs = new URLSearchParams({ stage_id: String(stageId), phase: String(phase), source: String(source) });
  const res = await fetch(`${apiBase()}/api/rollouts/${encodeURIComponent(rolloutId)}?${qs.toString()}`, { signal });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const data = await res.json();
  if (!data || typeof data !== "object") throw new Error("invalid response");
  if (!data.ok) throw new Error(data.error || "read rollout failed");
  return data;
}
